src/lib/project-saver-hoc.jsx

```diff
@@ -234,45 +233,14 @@ const ProjectSaverHOC = function (WrappedComponent) {
                     )
                 )
             ).then(() => {
-                const opts = {
-                    body: savedVMState,
-                    // If we set json:true then the body is double-stringified, so don't
-                    headers: {
-                        'Content-Type': 'application/json'
-                    },
-                    withCredentials: true
-                };
-                const creatingProject = projectId === null || typeof projectId === 'undefined';
-                let qs = queryString.stringify(requestParams);
-                if (qs) qs = `?${qs}`;
-                if (creatingProject) {
-                    Object.assign(opts, {
-                        method: 'post',
-                        url: `${storage.projectHost}/${qs}`
-                    });
-                } else {
-                    Object.assign(opts, {
-                        method: 'put',
-                        url: `${storage.projectHost}/${projectId}${qs}`
-                    });
+                const {onUpdateProjectData} = this.props;
+                const saveReturnValue = onUpdateProjectData(projectId,
+                    savedVMState, requestParams);
+                // Allow onUpdateProjectData to either return immediately or a promise.
+                if (!(saveReturnValue.then && saveReturnValue.catch)) {
```

Comment:
Can we rely on the supplier of `onUpdateProjectData` to ensure it returns a promise? If not, maybe we should wrap the entire thing in a promise that gets returned?  Something like:
```js
return new Promise((resolve, reject) => {
    try {
        const saveReturnValue = onUpdateProjectData(projectId, savedVMState, requestParams);
    } catch (e) {
        return reject(e);
    }
    // if saveReturnValue is a promise, this wrapper promise will resolve to what saveReturnValue resolves to
    return resolve(saveReturnValue); 
});
```

Reply:
Yeah i wasn't quite sure whether to enforce the promise requirement. That code block actually started as `if (!returnValue.then) { throw new Error('You must return a promise'); }` but then i realized i could just resolve at that point. Also you don't need to try/catch because it is already inside a promise chain, so throwing in that function would automatically go to the next `.catch` in the chain.

Reply:
Actually i think you can just `return Promise.resolve(onUpdateProjectData(...))` since promises unwrap the resolve value from other promises

Reply:
That looks good to me.

Reply:
Oh wait, even simpler, it is in a promise chain, it doesn't matter if it is a promise or not, either way the next then will get the same result at the right time